plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/MetadataLogEntriesTable.java

```diff
@@ -0,0 +1,136 @@
+import org.apache.iceberg.DataTask;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableScan;
+import org.apache.iceberg.io.CloseableIterable;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import java.util.Map;
+
+import static io.trino.plugin.iceberg.IcebergUtil.buildTableScan;
+import static io.trino.plugin.iceberg.IcebergUtil.columnNameToPositionInSchema;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.TimestampWithTimeZoneType.TIMESTAMP_TZ_MILLIS;
+import static io.trino.spi.type.Timestamps.MICROSECONDS_PER_MILLISECOND;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Objects.requireNonNull;
+import static org.apache.iceberg.MetadataTableType.METADATA_LOG_ENTRIES;
+
+public class MetadataLogEntriesTable
+        implements SystemTable
+{
+    private final ConnectorTableMetadata tableMetadata;
+    private final Table icebergTable;
+    private static final String TIMESTAMP_COLUMN_NAME = "timestamp";
+    private static final String FILE_COLUMN_NAME = "file";
+    private static final String LATEST_SNAPSHOT_ID_COLUMN_NAME = "latest_snapshot_id";
+    private static final String LATEST_SCHEMA_ID_COLUMN_NAME = "latest_schema_id";
+    private static final String LATEST_SEQUENCE_NUMBER_COLUMN_NAME = "latest_sequence_number";
+
+    public MetadataLogEntriesTable(SchemaTableName tableName, Table icebergTable)
+    {
+        this.icebergTable = requireNonNull(icebergTable, "icebergTable is null");
+
+        tableMetadata = new ConnectorTableMetadata(requireNonNull(tableName, "tableName is null"),
+                ImmutableList.<ColumnMetadata>builder()
+                        .add(new ColumnMetadata(TIMESTAMP_COLUMN_NAME, TIMESTAMP_TZ_MILLIS))
+                        .add(new ColumnMetadata(FILE_COLUMN_NAME, VARCHAR))
+                        .add(new ColumnMetadata(LATEST_SNAPSHOT_ID_COLUMN_NAME, BIGINT))
+                        .add(new ColumnMetadata(LATEST_SCHEMA_ID_COLUMN_NAME, INTEGER))
+                        .add(new ColumnMetadata(LATEST_SEQUENCE_NUMBER_COLUMN_NAME, BIGINT))
+                        .build());
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return Distribution.SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return tableMetadata;
+    }
+
+    @Override
+    public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        return new FixedPageSource(buildPages(tableMetadata, session, icebergTable));
+    }
+
+    private static List<Page> buildPages(ConnectorTableMetadata tableMetadata, ConnectorSession session, Table icebergTable)
```

Comment:
Same issue in Spark 3.3.x and Trino.
I don't know is this a correct behavior.

```
spark-sql> SELECT * FROM test.test_table.metadata_log_entries;
2024-01-22 16:50:33.208	hdfs://hadoop-master:9000/metadata/00000-acbc6857-46cd-498a-b8e0-95b5a8b661df.metadata.json	NULL	NULL	NULL
2024-01-22 16:50:39.595	hdfs://hadoop-master:9000/metadata/00001-9a9310e3-b54c-4bb1-a9ae-f8a775b51755.metadata.json	NULL	NULL	NULL
2024-01-22 16:50:40.7	hdfs://hadoop-master:9000/metadata/00002-72310e31-4d95-4f60-ba43-9f4e336f1ff9.metadata.json	NULL	NULL	NULL
2024-01-22 16:50:43.645	hdfs://hadoop-master:9000/metadata/00003-ef40bea2-3719-4174-890b-d3aced094ba4.metadata.json	NULL	NULL	NULL
2024-01-22 16:50:44.734	hdfs://hadoop-master:9000/metadata/00004-9698fb96-0ea7-48b4-a499-4ed6915a09b9.metadata.json	NULL	NULL	NULL
2024-01-22 16:51:02.057	hdfs://hadoop-master:9000/metadata/00005-679d6e5b-502d-4f93-a0d2-29d1de0ac01f.metadata.json	4266830935110946003	0	4
Time taken: 13.582 seconds, Fetched 6 row(s)


trino> SELECT * FROM ypoon."iceberg_trino_audit_log_public2$metadata_log_entries";
             timestamp              |                                                               file                                                               | latest_snapshot_id  | latest_schema_id | latest_sequence_number
------------------------------------+----------------------------------------------------------------------------------------------------------------------------------+---------------------+------------------+------------------------
 2024-01-22 16:50:33.208 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00000-acbc6857-46cd-498a-b8e0-95b5a8b661df.metadata.json |                NULL |             NULL |                   NULL
 2024-01-22 16:50:39.595 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00001-9a9310e3-b54c-4bb1-a9ae-f8a775b51755.metadata.json |                NULL |             NULL |                   NULL
 2024-01-22 16:50:40.700 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00002-72310e31-4d95-4f60-ba43-9f4e336f1ff9.metadata.json |                NULL |             NULL |                   NULL
 2024-01-22 16:50:43.645 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00003-ef40bea2-3719-4174-890b-d3aced094ba4.metadata.json |                NULL |             NULL |                   NULL
 2024-01-22 16:50:44.734 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00004-9698fb96-0ea7-48b4-a499-4ed6915a09b9.metadata.json |                NULL |             NULL |                   NULL
 2024-01-22 16:51:02.057 Asia/Tokyo | hdfs://hadoop-master:9000/metadata/00005-679d6e5b-502d-4f93-a0d2-29d1de0ac01f.metadata.json | 4266830935110946003 |                0 |                      4
(6 rows)
```

Reply:
The current implementation in Iceberg rely on `snapshot-log` field in metadata file:
https://github.com/apache/iceberg/blob/556b79893c4cf760f031258f0d8ae657e42b4443/core/src/main/java/org/apache/iceberg/MetadataLogEntriesTable.java#L110
https://github.com/apache/iceberg/blob/556b79893c4cf760f031258f0d8ae657e42b4443/core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java#L352
https://github.com/apache/iceberg/blob/556b79893c4cf760f031258f0d8ae657e42b4443/core/src/main/java/org/apache/iceberg/BaseTable.java#L150
`snapshot-log` got reset after after the CREATE OR REPLACE statement.
I can't find a precise definition for `latest_snapshot_id`, `latest_schema_id` and `latest_sequence_number`, so I can't verify is this a correct behavior.
All I know is the result should be the same across different engines since everyone is using `MetadataLogEntriesTable` from Iceberg lib.

Reply:
Spark is worse. More records become null. They don't set `snapshot-log` for CREATE and ALTER TABLE.
```
spark-sql> create table test.t1 (c1 integer);
Time taken: 0.27 seconds
spark-sql> alter table test.t1 add columns (c2 string);
Time taken: 0.242 seconds
spark-sql> SELECT * FROM test.t1.metadata_log_entries;
2024-01-22 18:51:41.331	hdfs://hadoop/tmp/test.db/t1/metadata/00000-d1ad9769-3bf3-497d-a86e-20f71d2700f7.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:41.593	hdfs://hadoop/tmp/test.db/t1/metadata/00001-9eb3275c-f53d-4e24-bc23-fdcadae28d92.metadata.json	NULL	NULL	NULL
Time taken: 1.212 seconds, Fetched 2 row(s)
spark-sql> insert into test.t1 values (1, 'a');
Time taken: 0.423 seconds
spark-sql> SELECT * FROM test.t1.metadata_log_entries;
2024-01-22 18:51:41.331	hdfs://hadoop/tmp/test.db/t1/metadata/00000-d1ad9769-3bf3-497d-a86e-20f71d2700f7.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:41.593	hdfs://hadoop/tmp/test.db/t1/metadata/00001-9eb3275c-f53d-4e24-bc23-fdcadae28d92.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:47.538	hdfs://hadoop/tmp/test.db/t1/metadata/00002-ec3f0aa6-7dae-4c45-90ad-13839eefbd54.metadata.json	1836642618692808023	1	0
Time taken: 0.14 seconds, Fetched 3 row(s)
spark-sql> delete from test.t1 where c1 = 1;
Time taken: 0.596 seconds
spark-sql> SELECT * FROM test.t1.metadata_log_entries;
2024-01-22 18:51:41.331	hdfs://hadoop/tmp/test.db/t1/metadata/00000-d1ad9769-3bf3-497d-a86e-20f71d2700f7.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:41.593	hdfs://hadoop/tmp/test.db/t1/metadata/00001-9eb3275c-f53d-4e24-bc23-fdcadae28d92.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:47.538	hdfs://hadoop/tmp/test.db/t1/metadata/00002-ec3f0aa6-7dae-4c45-90ad-13839eefbd54.metadata.json	1836642618692808023	1	0
2024-01-22 18:51:52.6	hdfs://hadoop/tmp/test.db/t1/metadata/00003-806771f6-e7f1-44f5-ac80-350f2c084505.metadata.json	8876099574020403871	1	0
Time taken: 0.121 seconds, Fetched 4 row(s)
spark-sql> CALL local.system.rewrite_data_files('test.t1');
0	0	0
Time taken: 0.173 seconds, Fetched 1 row(s)
spark-sql> SELECT * FROM test.t1.metadata_log_entries;
2024-01-22 18:51:41.331	hdfs://hadoop/tmp/test.db/t1/metadata/00000-d1ad9769-3bf3-497d-a86e-20f71d2700f7.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:41.593	hdfs://hadoop/tmp/test.db/t1/metadata/00001-9eb3275c-f53d-4e24-bc23-fdcadae28d92.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:47.538	hdfs://hadoop/tmp/test.db/t1/metadata/00002-ec3f0aa6-7dae-4c45-90ad-13839eefbd54.metadata.json	1836642618692808023	1	0
2024-01-22 18:51:52.6	hdfs://hadoop/tmp/test.db/t1/metadata/00003-806771f6-e7f1-44f5-ac80-350f2c084505.metadata.json	8876099574020403871	1	0
Time taken: 0.106 seconds, Fetched 4 row(s)
spark-sql> create or replace table test.t1 (c3 integer);
Time taken: 0.236 seconds
spark-sql> SELECT * FROM test.t1.metadata_log_entries;
2024-01-22 18:51:41.331	hdfs://hadoop/tmp/test.db/t1/metadata/00000-d1ad9769-3bf3-497d-a86e-20f71d2700f7.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:41.593	hdfs://hadoop/tmp/test.db/t1/metadata/00001-9eb3275c-f53d-4e24-bc23-fdcadae28d92.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:47.538	hdfs://hadoop/tmp/test.db/t1/metadata/00002-ec3f0aa6-7dae-4c45-90ad-13839eefbd54.metadata.json	NULL	NULL	NULL
2024-01-22 18:51:52.6	hdfs://hadoop/tmp/test.db/t1/metadata/00003-806771f6-e7f1-44f5-ac80-350f2c084505.metadata.json	NULL	NULL	NULL
2024-01-22 18:52:01.749	hdfs://hadoop/tmp/test.db/t1/metadata/00004-d9c74f49-37c8-4463-8565-7b31660723e2.metadata.json	NULL	NULL	NULL
Time taken: 0.159 seconds, Fetched 5 row(s)
```

Reply:
@oneonestar what do you plan to do about this bug?
Did you discover the cause of it ? Is it related to the `apache/iceberg` library? If yes, can you pls file an issue about it?
Let's also "document" the erroneous behavior in a test (to know also when it is going to be fixed).

Reply:
Asking for a precise definition for `latest_snapshot_id, latest_schema_id, latest_sequence_number` https://github.com/apache/iceberg/issues/9723

Reply:
Added a test for this.
This is going to be fixed in Iceberg 1.5.0.